pkg/gui/gui.go

```diff
@@ -117,46 +137,91 @@ func NewGui(log *logrus.Entry, gitCommand *commands.GitCommand, oSCommand *comma
 	return gui, nil
 }
 
-func (gui *Gui) scrollUpMain(g *gocui.Gui, v *gocui.View) error {
-	mainView, _ := g.View("main")
-	ox, oy := mainView.Origin()
-	if oy >= 1 {
-		return mainView.SetOrigin(ox, oy-gui.Config.GetUserConfig().GetInt("gui.scrollHeight"))
+// Run setup the gui with keybindings and start the mainloop.
+// returns an error if something goes wrong.
+func (gui *Gui) Run() error {
+
+	var err error
+
+	gui.g, err = gocui.NewGui(gocui.OutputNormal, OverlappingEdges)
+	if err != nil {
+		gui.Log.Errorf("Failed at newgui: %s\n", err)
+		return err
 	}
-	return nil
-}
 
-func (gui *Gui) scrollDownMain(g *gocui.Gui, v *gocui.View) error {
-	mainView, _ := g.View("main")
-	ox, oy := mainView.Origin()
-	if oy < len(mainView.BufferLines()) {
-		return mainView.SetOrigin(ox, oy+gui.Config.GetUserConfig().GetInt("gui.scrollHeight"))
+	defer gui.g.Close()
+
+	err = gui.SetColorScheme()
+	if err != nil {
+		gui.Log.Errorf("Failed at setcolorscheme: %s\n", err)
+		return err
+	}
+
+	gui.g.SetManagerFunc(gui.layout)
+
+	gui.goEvery(time.Second*60, gui.fetch)
+	gui.goEvery(time.Second*10, gui.refreshFiles)
+	gui.goEvery(time.Millisecond*50, gui.updateLoader)
+	gui.goEvery(time.Millisecond*50, gui.renderAppStatus)
+
+	if err = gui.keybindings(gui.g); err != nil {
+		gui.Log.Errorf("Failed to set keybindings at Run: %s\n", err)
+		return err
+	}
+
+	err = gui.g.MainLoop()
+	if err != nil {
+		gui.Log.Errorf("Failed to run mainLoop at Run: %s\n", err)
+		return err
 	}
+
 	return nil
 }
 
-func (gui *Gui) handleRefresh(g *gocui.Gui, v *gocui.View) error {
-	return gui.refreshSidePanels(g)
-}
+// RunWithSubprocesses loops, instantiating a new gocui.Gui with each iteration
+// if the error returned from a run is a ErrSubProcess, it runs the subprocess
+// otherwise it handles the error, possibly by quitting the application
+func (gui *Gui) RunWithSubprocesses() {
+	for {
+
+		err := gui.Run()
+		if err != nil {
+			if err == gocui.ErrQuit {
+				break
+			} else if err == gui.Errors.ErrSubProcess {
+				gui.SubProcess.Stdin = os.Stdin
+				gui.SubProcess.Stdout = os.Stdout
+				gui.SubProcess.Stderr = os.Stderr
+
+				err = gui.SubProcess.Run()
+				if err != nil {
+					gui.Log.Errorf("Failed to runWithSubProcess: %s\n", err)
+					return
```

Comment:
Looks like we should either panic here or change the function to return an error

Reply:
Added a panic instead of a return